Fixed Tool Tips Rendering Within the Window on Second Monitors

rainbain opened the pull request:
General improvements, mainly to the bubble tool tips, to make them feel less in the way and more clear. 

 * Bubble tool tip arrow offset to point at the option referenced so that it's easier to tell what the tool tip is for when clipped by the edges of the screen.
* Offset bubble tool tip above and below the control point so the arrow does not cover the option.
* 750 ms instead of 300 ms for the tool tip so it does not get in the way as you read through the settings. 
 * 3px bubble tool tip corners instead of 7px corners to be more visually consistent with the square corners of the other tool tips.

MayImilae commented:
> 750 ms instead of 300 ms for the tool tip so it does not get in the way as you read through the settings.

That was intentional so users can get the information without a bunch of delay as they scan through the settings. It's a trade off, and I think we made the correct choices for the duration. That shouldn't be changed.

There are other ways to keep it out of the way, such as horizontal tooltips. But that's probably beyond the scope of this PR.

rainbain commented:
Alright awesome, thank you for the reply.

After work today I am going to put back the borders and delay. Before I made this I was weighing the tradeoffs and that seems very reasonable.

Now that you mention that they should already try and stay attached to message, I am going to want to go back and take a closer look at why exactly they are not doing this. After looking at your image, it does that on one of my monitors and does not do that on my other monitor.
![Screenshot_20240610_142342](https://github.com/dolphin-emu/dolphin/assets/67609611/7d998228-a9b8-4645-952a-c7a5f33d036f)
I am pretty sure what is already responsible for this now so it should be a lot easier than offsetting them.

MayImilae commented:
We have had a lot of issues with the tooltips on secondary monitors. See #12650. If you can figure it out that would be awesome!

rainbain commented:
Alright I reworked it a bit. I put back the 750ms delay and the 7px corners and made it consistent across monitors.

The issue seems to be that it was referencing screen space. On the first screen the left would be 0px and it would push the tool tip now at negative coordinates to 0. Putting it at 0 made it spawn at the left most border of the config window. This would break if it was on a second monitor though, where left was not 0px.

I made it so the widget references the parent window's geometry and now the tool tips stay within the dolphin window.
A similar change was done for when determining to spawn the tool tip on the bottom. If the y pos was bellow 0 it would go on the bottom, so it would only clip if it was on the top most monitor and was already outside the dolphin window. Now it goes on the bottom if it goes outside the window geometry.

After changing how the tool tips were placed, the arrow did not always perfectly point at the control point, so I added the dynamic offset agin to make them consistently point at the control point.

Now it seems the tool tips spawn the same as they would on the first monitor, but on other monitors! Let me know if this fix seems correct. There is room for interpretation on how they should spawn.

rainbain commented:
Alright that should be all the errors fixed. I will need to change my build environment to test that on my side so I can check easily if something is up.

mbc07 commented:
@dolphin-emu-bot rebuild

mbc07 commented:
@dolphin-emu-bot rebuild

AdmiralCurtiss commented:
Can you edit the PR description at the top so it matches what's now in the PR?

Also please rebase & squash your changes on top of master.

BhaaLseN commented:
I just ran into the alignment issue on my secondary screen (where the tooltip would extend to the _left_ rather than to the _right_) and can confirm this branch fixes this. 👍 

Didn't review the code (yet) though.

rainbain commented:
Yes the rebase did go quite wrong, I do not use git every day, so I am always a bit rusty with it. Anyhow that is fine, this should of reset it. I also went ahead and fixed a bug I found. I tried this code on another PC with 2 screens, and it seems the window system of that PC would return bad geometry some of the time. I changed it out to use mapToGlobal to get the window geometry.

This should be better now! Let me know what you think.

Dentomologist reviewed (changes requested):
Offsetting the BalloonTip for ComboBoxes or other controls that might have text obscured by the arrow tip is a good idea. We may need to account for DPI settings instead of just hardcoding 10, but I haven't tested that yet.

I don't think we should try to restrict tooltips to their generating window. Tooltips are supposed to pop out, and every other application I checked allows them to extend out of the window.

This is particularly true when the window is partially off the screen and the visible portion is too small for the tooltip. As it stands this PR can have most of the BalloonTip being off-screen and unreadable, which is a bigger problem than the arrow tip not pointing to the ideal place.

We can fix the balloon directions on secondary displays by changing [these](https://github.com/dolphin-emu/dolphin/blob/5af0ae25e6ebeaec187950f579f722637546740f/Source/Core/DolphinQt/Config/ToolTipControls/BalloonTip.cpp#L228-L231) lines to:
```suggestion
  const bool arrow_at_bottom =
      target_arrow_tip_position.y() - 10 - size_hint.height() + arrow_height >= screen_rect.top();
  const bool arrow_at_left =
      target_arrow_tip_position.x() + size_hint.width() - arrow_tip_x_offset <=
      screen_rect.left() + screen_rect.width();
```
... where the 10 for arrow_at_bottom is the magnitude of tip_vertical_offset.

I think fixing the arrow-in-the-wrong-place issue when the target is close to the edge of the screen (while keeping the whole balloon on-screen) is going to require changing how the BalloonTip is drawn in that case. I have an idea for that I've had on the back burner.
